File: librarian-mcp/stitchpunks/sweeper/sweeper_query.ts
/**
 * Sweeper Scribe — MCP Tool: sweeper_query
 * KN016 / A&A #2328 candidate
 *
 * MCP tool for Founder query: "what's stale?" / "what's orphaned?" / "what's drifted?"
 * Returns structured digest from the most recent scan log.
 *
 * Toolsmith log: TS-BISHOP-SWEEPER-SCAVENGER-KN016-BP002
 */

import * as fs from "fs";
import * as path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DIGEST_LOG_PATH = path.join(__dirname, "digest_log.jsonl");

export interface SweeperDigestEntry {
  lines: string[];
  chronos_hash: string;
  timestamp: string;
  total_items: number;
}

export interface SweeperQueryResult {
  query_class: string;
  latest_digest: SweeperDigestEntry | null;
  all_digests_count: number;
  formatted: string;
}

/**
 * Load the most recent digest from the log.
 */
function loadLatestDigest(): SweeperDigestEntry | null {
  if (!fs.existsSync(DIGEST_LOG_PATH)) {
    return null;
  }
  const lines = fs
    .readFileSync(DIGEST_LOG_PATH, "utf-8")
    .split("\n")
    .filter((l) => l.trim());
  if (lines.length === 0) return null;
  try {
    return JSON.parse(lines[lines.length - 1]) as SweeperDigestEntry;
  } catch {
    return null;
  }
}

/**
 * Count all digest entries in the log.
 */
function countDigests(): number {
  if (!fs.existsSync(DIGEST_LOG_PATH)) return 0;
  return fs
    .readFileSync(DIGEST_LOG_PATH, "utf-8")
    .split("\n")
    .filter((l) => l.trim()).length;
}

/**
 * Main query function.
 * query_class: "stale" | "orphaned" | "drift" | "all"
 */
export function sweeperQuery(queryClass: string = "all"): SweeperQueryResult {
  const latest = loadLatestDigest();
  const count = countDigests();

  const formatted = latest
    ? latest.lines.join("\n")
    : "No sweep digest available — run sweeper scan first.";

  return {
    query_class: queryClass,
    latest_digest: latest,
    all_digests_count: count,
    formatted,
  };
}

// MCP tool entrypoint
if (process.argv[2] === "--query") {
  const cls = process.argv[3] || "all";
  const result = sweeperQuery(cls);
  console.log(JSON.stringify(result, null, 2));
}

File: librarian-mcp/stitchpunks/sweeper/sweeper_query.py
"""
Sweeper Scribe — Python MCP Query Wrapper
KN016 / A&A #2328 candidate

Python-accessible interface to sweeperQuery, wrapping the digest log.
Used by tests_kn016.py and by direct MCP invocation.

Toolsmith log: TS-BISHOP-SWEEPER-SCAVENGER-KN016-BP002
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, Optional

_HERE = Path(__file__).parent
DIGEST_LOG_PATH = _HERE / "digest_log.jsonl"


def _load_latest_digest(log_path: Optional[Path] = None) -> Optional[Dict[str, Any]]:
    path = log_path or DIGEST_LOG_PATH
    if not path.exists():
        return None
    lines = [l.strip() for l in path.read_text(encoding="utf-8").split("\n") if l.strip()]
    if not lines:
        return None
    try:
        return json.loads(lines[-1])
    except (json.JSONDecodeError, IndexError):
        return None


def _count_digests(log_path: Optional[Path] = None) -> int:
    path = log_path or DIGEST_LOG_PATH
    if not path.exists():
        return 0
    return len([l for l in path.read_text(encoding="utf-8").split("\n") if l.strip()])


def sweeperQuery(
    query_class: str = "all",
    log_path: Optional[Path] = None,
) -> Dict[str, Any]:
    """
    Query the sweeper digest log.

    query_class: "stale" | "orphaned" | "drift" | "all"
    Returns: { query_class, latest_digest, all_digests_count, formatted }
    """
    latest = _load_latest_digest(log_path)
    count = _count_digests(log_path)

    formatted = (
        "\n".join(latest["lines"]) if latest
        else "No sweep digest available — run sweeper scan first."
    )

    return {
        "query_class": query_class,
        "latest_digest": latest,
        "all_digests_count": count,
        "formatted": formatted,
    }
